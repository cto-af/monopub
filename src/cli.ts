import {Command, Option, type OutputConfiguration} from 'commander';
import {MonoRoot} from './index.js';
import path from 'node:path';
import {version} from './version.js';

function push(p: string, prev: string[] = []): string[] {
  prev.push(p);
  return prev;
}

/**
 * Run the command line interface for a given set of arguments.
 *
 * @param args Arguments to process, in node format.  If not specified, uses
 *   process.argv.
 * @param out Override stdout and stderr for testing.
 */
export async function cli(
  args?: string[],
  out?: OutputConfiguration
): Promise<void> {
  const program = new Command();
  if (out) {
    program.configureOutput(out);
    program.exitOverride();
  }

  program
    .option('-p, --packages <glob>', 'Glob to find names of sub-package directories.  Can be specified multiple times.', push, [])
    .option('-P, --private', 'Include private packages')
    .addOption(
      new Option('--cwd <directory>', 'Operate from given directory')
        .default(process.cwd(), 'cwd')
        .argParser(val => path.resolve(process.cwd(), val))
    )
    .version(version)
    .configureHelp({
      showGlobalOptions: true,
      sortOptions: true,
      sortSubcommands: true,
    });

  program
    .command('version')
    .description('Update version of all sub-repos to match the root.  Should be called from the npm "version" script.')
    .action(async(_opts, p) => {
      const opts = p.optsWithGlobals();
      const m = await new MonoRoot(opts).init();
      m.setVersions();
      await m.save(true);
    });

  program
    .command('delete <fields>')
    .description('Delete this set of comma-separated fields from all package files.')
    .action(async(fields, _opts, p) => {
      const opts = p.optsWithGlobals();
      const m = await new MonoRoot(opts).init();
      m.delete(fields.split(','));
      await m.save(false);
    });

  program
    .command('exec <command>')
    .description('Run the given command in a shell for each package file.')
    .action(async(cmd, _opts, p) => {
      const opts = p.optsWithGlobals();
      const m = await new MonoRoot(opts).init();
      const results = await m.execAll({stdio: 'inherit'}, cmd);
      if (results.some(r => !r.ok)) {
        program.error('Failed execution');
      }
    });

  await program.parseAsync(args);
}
