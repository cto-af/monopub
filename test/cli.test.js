import assert from 'node:assert';
import {cli} from '../lib/cli.js';
import {fileURLToPath} from 'node:url';
import snap from 'snappy-snaps';
import test from 'node:test';

const bin = fileURLToPath(new URL('../bin/monopub.js', import.meta.url));
const cwd = fileURLToPath(new URL('fixtures', import.meta.url));

async function runCli(...args) {
  let stdout = '';
  let stderr = '';
  let error = null;
  try {
    await cli([process.argv0, bin, ...args], {
      writeOut: str => (stdout += str),
      writeErr: str => (stderr += str),
    });
  } catch (er) {
    error = er;
  }
  return {stdout, stderr, code: error?.code};
}

test('help', async() => {
  const base = await runCli('-h');
  assert.deepEqual(base, await snap('base', base));

  const del = await runCli('delete', '-h');
  assert.deepEqual(del, await snap('delete', del));

  const exec = await runCli('exec', '-h');
  assert.deepEqual(exec, await snap('exec', exec));

  const version = await runCli('version', '-h');
  assert.deepEqual(version, await snap('version', version));
});

test('exec', async() => {
  const fixtures = await runCli(
    '--cwd',
    cwd,
    '-P',
    '-p',
    'foo',
    '-p',
    'bar',
    'exec',
    'exit 0'
  );
  assert.equal(fixtures.code, undefined);

  const fail = await runCli(
    '--cwd',
    cwd,
    '-P',
    '-p',
    'foo',
    '-p',
    'bar',
    'exec',
    'exit 1'
  );
  assert.deepEqual(fail, {
    stdout: '',
    stderr: 'Failed execution\n',
    code: 'commander.error',
  });
});

test('version', async() => {
  const version = await runCli(
    '--cwd',
    cwd,
    '-P',
    '-p',
    'foo',
    '-p',
    'bar',
    'version'
  );
  assert.equal(version.code, undefined);
});
