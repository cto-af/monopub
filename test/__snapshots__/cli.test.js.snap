// Data Snap v1

exports[`base`] = {
  "data": {
    "stdout": "Usage: monopub [options] [command]\n\nOptions:\n  --cwd \u003Cdirectory\u003E      Operate from given directory (default: cwd)\n  -h, --help             display help for command\n  -p, --packages \u003Cglob\u003E  Glob to find names of sub-package directories.  Can be\n                         specified multiple times. (default: [])\n  -P, --private          Include private packages\n  -V, --version          output the version number\n\nCommands:\n  delete \u003Cfields\u003E        Delete this set of comma-separated fields from all\n                         package files.\n  exec \u003Ccommand\u003E         Run the given command in a shell for each package\n                         file.\n  help [command]         display help for command\n  order                  Output the order in which packages will be processed.\n  version                Update version of all sub-repos to match the root.\n                         Should be called from the npm \"version\" script.\n",
    "stderr": "",
    "code": "commander.helpDisplayed"
  },
  "meta": {
    "expires": null
  }
}

exports[`delete`] = {
  "data": {
    "stdout": "Usage: monopub delete [options] \u003Cfields\u003E\n\nDelete this set of comma-separated fields from all package files.\n\nOptions:\n  -h, --help             display help for command\n\nGlobal Options:\n  --cwd \u003Cdirectory\u003E      Operate from given directory (default: cwd)\n  -p, --packages \u003Cglob\u003E  Glob to find names of sub-package directories.  Can be\n                         specified multiple times. (default: [])\n  -P, --private          Include private packages\n  -V, --version          output the version number\n",
    "stderr": "",
    "code": "commander.helpDisplayed"
  },
  "meta": {
    "expires": null
  }
}

exports[`exec`] = {
  "data": {
    "stdout": "Usage: monopub exec [options] \u003Ccommand\u003E\n\nRun the given command in a shell for each package file.\n\nOptions:\n  -h, --help             display help for command\n\nGlobal Options:\n  --cwd \u003Cdirectory\u003E      Operate from given directory (default: cwd)\n  -p, --packages \u003Cglob\u003E  Glob to find names of sub-package directories.  Can be\n                         specified multiple times. (default: [])\n  -P, --private          Include private packages\n  -V, --version          output the version number\n",
    "stderr": "",
    "code": "commander.helpDisplayed"
  },
  "meta": {
    "expires": null
  }
}

exports[`version`] = {
  "data": {
    "stdout": "Usage: monopub version [options]\n\nUpdate version of all sub-repos to match the root.  Should be called from the\nnpm \"version\" script.\n\nOptions:\n  -h, --help             display help for command\n\nGlobal Options:\n  --cwd \u003Cdirectory\u003E      Operate from given directory (default: cwd)\n  -p, --packages \u003Cglob\u003E  Glob to find names of sub-package directories.  Can be\n                         specified multiple times. (default: [])\n  -P, --private          Include private packages\n  -V, --version          output the version number\n",
    "stderr": "",
    "code": "commander.helpDisplayed"
  },
  "meta": {
    "expires": null
  }
}
